tools/cell_census_builder/__main__.py

```diff
@@ -210,78 +200,133 @@ def build_step1_get_source_assets(args: argparse.Namespace, assets_path: str) ->
     return datasets
 
 
-def build_step2_create_axis(
-    soma_path: str,
-    assets_path: str,
-    datasets: List[Dataset],
-    experiment_builders: List[ExperimentBuilder],
-    args: argparse.Namespace,
-) -> Tuple[soma.Collection, List[Dataset]]:
-    """
-    Create all objects, and populate the axis dataframes.
-
-    Returns: the filtered datasets that will be included. This is simply
-    an optimization to allow subsequent X matrix writing to skip unused
-    datasets.
-    """
-    logging.info("Build step 2 - axis creation - started")
-
-    top_level_collection = create_top_level_collections(soma_path)
-
-    # Create axis
-    for e in experiment_builders:
-        e.create(data_collection=top_level_collection[CENSUS_DATA_NAME])
-        assert soma.Experiment(e.se_uri).exists()
-
-    # Write obs axis and accumulate var axis (and remember the datasets that pass our filter)
+def populate_obs_axis(
+    assets_path: str, datasets: List[Dataset], experiment_builders: List[ExperimentBuilder]
+) -> List[Dataset]:
     filtered_datasets = []
     N = len(datasets) * len(experiment_builders)
-    n = 1
+    n = 0
+
     for dataset, ad in open_anndata(assets_path, datasets, backed="r"):
         dataset_total_cell_count = 0
-        for e in experiment_builders:
-            dataset_total_cell_count += e.accumulate_axes(dataset, ad, progress=(n, N))
+
+        for eb in reopen_experiment_builders(experiment_builders):
             n += 1
+            logging.info(f"{eb.name}: filtering dataset '{dataset.dataset_id}' ({n} of {N})")
+            ad_filtered = eb.filter_anndata_cells(ad)
 
-        dataset.dataset_total_cell_count = dataset_total_cell_count
+            if len(ad_filtered.obs) == 0:
+                logging.info(f"{eb.name} - H5AD has no data after filtering, skipping {dataset.dataset_h5ad_path}")
+                continue
+
+            # append to `obs`; accumulate `var` data
+            dataset_total_cell_count += eb.accumulate_axes(dataset, ad_filtered)
+
+        # dataset passes filter if either experiment includes cells from the dataset
         if dataset_total_cell_count > 0:
             filtered_datasets.append(dataset)
+            dataset.dataset_total_cell_count = dataset_total_cell_count
+
+    for eb in experiment_builders:
+        logging.info(f"Experiment {eb.name} will contain {eb.n_obs} cells from {eb.n_datasets} datasets")
+
+
```

Comment:
formatting issues here (too many lines).

I'm thinking the CI linter is not working for the builder, or it would have caught this....

Reply:
We don't need no stinkin' linter--we've got Bruce! ;-) I think all is well wrt the linter checks. Linter not running due to conflict w/main.

Reply:
doh!  of course.  thank you...